notebook.md

## Notes

This notebook was began after much of the development of the ChemoR pipeline had been completed and when a significant re-write was undertaken to begin to prepare for publication. The following contains a list of things that needs to be done, and the approaches attempted to achieve these goals.

## Tasks

### Organization

1. Begin to organize with the "\_N" nomenclature to keep output files organized
2. Add better comments in the .sh scripts
3. Rewrite in Nextflow


### Improvements

1. Remove all reliance on the `-max_target_seqs` option in `blastp` commands
2. Improve the final reciprocal BLAST filtering strategy
  - It currently relies upon a `grep` comparison with the output from the script on the *C. elegans* genome
  - It should rely upon a list of hard-coded, curated genes
  - It also removes putative ChemoRs that **do not** have a hit to a *C. elegans* gene, which is a bug
  - It should keep those putative ChemoRs and even focus on them
3. Add a search using putative ChemoRs against the genome that they come from (to get diverged sequences)
4. Be explicit about Gene_ID and Transcript_ID, and make a master data frame with everything
5. TM prediction on everything prior to phylogenetic inference
6. Change what species go into the tree
7. When there are isoforms, keep the longest sequence


### Rolling Notebook

#### 2019-05-02

- began this notebook  
- met with Mostafa to plan ahead for publication/preprint  
- updated `WBP_download.sh` to not discriminate between genome "quality" and include everything in the same `genomes/` directory in `GHdata`
- ran `sh WBP_download.sh` to update the genomes
- deleted `g_genomes`, `ngf_genomes`, `f_genomes`, `ngo_genomes`, `a_genomes`, and `p_genomes`
- keep `7tm_1` and ChemoR hits in the first `hmmsearch` and the reciprocal search
- on line 164 in `ChemoR_analysis.sh`, the bit where the reciprocal BLAST output is compared to *C. elegans* ChemoRs


#### 2019-05-03

Looking at the removed IDs listed in `_3_filtered.txt`, it appears that there are problems with the comparison to putative *C. elegans* ChemoRs. This list was generated by probing **proteins** that had been labeled with any of the 23 family names, but it is clear there are other sequences in WormBase that should be annotated as ChemoRs but aren't. For instance "Y40B10A.5a" is a ChemoR (contains a srsx domain), but it isn't on the list I curated and it doesn't have a protein name. This is the top hit to "Bm10687", so that gene is incorrectly removed.

Possible solutions:

- **have a more complete curated list from *C. elegans***
- compare to a different list entirely - originally I compared to the output of the pipeline on the *C. elegans* genome - e.g. `caenorhabditis_elegans_2_ids.txt`
  - if I reimplemented this approach, it would correctly keep "Bm10687" (true positive) but incorrectly keep "Bm1100a" (false positive)
  - in previous iterations, I assumed false positive were peptide receptors that would cluster with srw family receptors, so I rooted on srw and made caveats to inferring anything from those gene counts
- I could keep any hit that had a *C. elegans* ChemoR in the top N hits
  - probably wouldn't work; "Bm10687" has two isoforms of a non-annotated ChemoR as the only hits, and these show weak similarity (e-value = 0.006 and 0.009)
- don't rely on the *C. elegans* filter
  - but the filter kind of works - it correctly removes "Bm1100a" because its top hit is "H22D07.1", or "gnrr-5"

Ways to get a better curated list:

- keep Compara gene trees of all the proteins I currently have
  - will require learning the WBP API
  - test: does "Y40B10A.5a" appear in the gene tree of one of the proteins I have already selected
  - answer: yes
- wget command to get a `.json` file of *C. elegans* paralogues for a specific Gene_ID:
    `wget -q --header='Content-type:application/json' 'https://parasite.wormbase.org/rest-13/homology/symbol/caenorhabditis_elegans_PRJNA13758/WBGene00005027?format=condensed;type=paralogues'  -O sra-1_paralogues.json`
- turned this into a Python script: `json_parser.py`
- de-duplicated *C. elegans* ChemoR sequences are at `"~/GitHub/50HGI/auxillary/ChemoR/simplemine_results.txt`
- **confirm the maintenance of true positives**

#### 2019-05-06

- began the Phylognetics block of `ChemoR_analysis.sh`
- edited `id_change.py` so that I don't have to rename files after it ran
- realized that I missed the srj family when curating the stock *C. elegans* ChemoR list, so added those to `celegans_chemor.csv` and repopulated `celegans_chemor_paralogues_geneid.txt` and `simplemine_results.txt`
- need to fix the 3rd filter so that genes without a significant hit to *C. elegans* ChemoRs are **not** removed
  - for example, "Bm17545" doesn't have a significant hit but it is likely a srsx receptor
  - **fixed**

#### 2019-05-07s

- seems to be some issues with the default `sed` used across computers
  - gnu-sed may be used on the iMac at the lab
    - yes, that is the issue (using `gsed` worked)
- `mafft` comments:
      To keep the alignment length, 42305 letters were DELETED.
      To know the positions of deleted letters, rerun the same command with the --mapout option.

      Strategy:
      Multi-INS-full (Not tested.)
      ?

      If unsure which option to use, try 'mafft --auto input > output'.
      For more information, see 'mafft --help', 'mafft --man' and the mafft page.

      The default gap scoring scheme has been changed in version 7.110 (2013 Oct).
      It tends to insert more gaps into gap-rich regions than previous versions.
      To disable this change, add the --leavegappyregion option.
- began tree inference using IQ-TREE and RAxML-NG
  - IQ-TREE command:
    `iqtree -s ../4/ds_reps_3.aln -nt 4 -alrt 1000 -bb 1000`
    -Options:
      `-s` = path to alignment file
      `-nt` = number of CPUs
      `-alrt` = run the SH-aLRT ultrafast bootstrap (approximate likelihood-ratio test)
      `-bb` = run 1000 ultrafast bootstraps (which are compared by the aLRT)
  - RAxML-NG commands:
    `# check alignment`
    `raxml-ng --parse --msa ../4/ds_reps_3.aln --model VT+G --prefix tree_1`
      `--parse` = check alignment for errors and compress MSA to binary
      `--model VT+G` = VT model with gamma distribution of rate heterogeneity
    `# run ML inference`
    `raxml-ng --all --msa tree_1.raxml.rba --model VT+G --prefix tree_1 --seed 2 --threads 5 --bs-metric fbp,tbe`
      `--seed` = set seed for reproducibility
      `--bs-metric` = use FBP (Felsenstein boostrap proportions) and TBE (transfer bootstrap expectation) metrics for compsarison of bootstrapped trees

#### 2019-05-08

- both IQ-TREE and RAxML-NG are running
- IQ-TREE selected the VT+F+R10 model
  - General ‘Variable Time’ matrix (Muller and Vingron, 2004)
  - Empirical base frequencies
  - FreeRate model; not necessarily gamma-distributed, split into 10 categories
- it's good that the model selected by IQ-TREE is very similar to the model I chose for RAxML-NG

#### 2019-05-15

- RAxML-ng was taking too long, so I quit it
- using the IQTREE output going forward
- after finishing tree annotation, I ensured the output of `tree_clades.csv` included **all** the original taxa
- next I will make a BLAST database from this annotated file, and BLAST all sequences not included in the tree against that DB

#### 2019-06-05

- made blast database of iqtree/tree_clades.csv

#### 2019-06-06

- BLASTP against tree DB
- a problem: awhile back I decided to label sequences with 4 characters of genus+species instead of 5, and I updated `id_change.py` accordingly. I did not realize that "acan" would then be the abbreviation for two species. I now need to create a liftover file that will replace all instances of the 4 character label with the original 5 character label (excluding Loa loa), and for A. caninum and A. cantonensis, this will need to incorporate the Transcript_IDs
  - I think I will manually replace the "acan" ID's in `ds_reps_3.aln.contree`, regenerate `tree_clades.csv`, and update `species_info.csv` and `clade_species_phylum.csv` accordingly
  - I also have to update the label FASTA files for these two species
    - NOTE: id_change.py will not be consistent with my actual labeled FASTA
